cde/doc/cde1/man/man1/dtprintinfo.1
.de LI
.\" simulate -mm .LIs by turning them into .TPs
.TP \\n()Jn
\\$1
..
.TH dtprintinfo 1 "12 October 1994"
.\" (c) Copyright 1993, 1994 Hewlett-Packard Company
.\" (c) Copyright 1993, 1994 International Business Machines Corp.
.\" (c) Copyright 1993, 1994 Sun Microsystems, Inc.
.\" (c) Copyright 1993, 1994 Novell, Inc.
.SH NAME
dtprintinfo \- the Common Desktop Environment Print Viewer
.SH SYNOPSIS
.yS
dtprintinfo 
[ -p <\f2printer\fP> | -all | -populate | -help ]
.yE
.SH DESCRIPTION
The dtprintinfo program provides a graphical interface that
displays the status of print queues and print jobs.
Additional information about print queues or print
jobs can be retrieved within the interface, individual print
queue labels and icons can be customized, and individual
print jobs can be canceled.  Print jobs can be submitted by
dragging files from the File Manager utility to a print queue
icon in a dtprintinfo window.
.SS KEY SUPPORTED TASKS
.IP \(em 3
Displaying configured print queues on the system.
.IP \(em 3
Displaying print queue and print job properties.
.IP \(em 3
Changing print queue labels and icons.
.IP \(em 3
Submitting print jobs.
.IP \(em 3
Canceling print jobs.
.SH NO OPTIONS
.sp .5
When the dtprintinfo command is started without any options, 
it shows the LPDEST printer or the system default printer if
LPDEST is not set.
.SH OPTIONS
.sp .5
.SS -p \f2<printer>\fP
Display information only for the specified \f2<printer>\fP.
.SS -all
Display information for all printers.
.SS -populate
The
.I -populate
option is available only for root. This
option creates default printer actions for all printers
in /etc/dt/appconfig/types/LANG. If LANG is not defined, 
the actions are created in /etc/dt/appconfig/types/C.
The post-installation process uses the 
.I -populate
feature to create default printer actions in /etc/dt/appconfig/types/C.
.SS -help
Display command line option help.
.SH RESOURCES
Resource are stored in /usr/dt/app-defaults/$LANG/Dtprintinfo.
.sp .5
.SS \f2updateInterval\fP
The \f2updateInterval\fP resource determines how often the dtprintinfo
program checks the printer.  The default value is every 30 seconds.
.SH ENVIRONMENT
.sp .5
.SS LPDEST
This environment variable determines which print queue to display 
when the dtprintinfo command is started without any options.
.SH BROADCAST MESSAGES
None.
.SH WARNING MESSAGES
.sp .5
.RS
.P
- Are you sure you want to cancel <\f2job\fP>? -
.RE
The dtprintinfo program displays a prompt before a print <\f2job\fP> is canceled.
.SH "EXIT STATUS"
The following exit values are returned:
.PP
.RS 3
.nr )J 4
.LI \h"\w'>'u"0
Successful completion.
.LI >0
An invocation error was detected.
.PP
.RE
.nr )J 0
.SH EXAMPLES
.SS dtprintinfo
.sp .5
Display the status of the LPDEST printer or the system default printer.
.SS dtprintinfo -p lp0
.sp .5
Display the status of the print queue called \f2lp0\fP.
.SS dtprintinfo -all
.sp .5
Display the status of all print queues.
.SH FILES
.sp .5
The dtprintinfo program stores a <printer>_Print.dt datatype file for
each customized print queue action in the $HOME/.dt/types directory.
.SH "SEE ALSO"
.BR dtlp (1), dtprintinfoaction (5)


cde/cde1/dtprintinfo/DtPrintinfo.C
// User Interface headers
#include "DtApp.h"
#include "DtPrinterIcon.h"
#include "Invoke.h"

#ifndef NO_CDE
extern "C" {
  #include <Dt/EnvControl.h>
}
#endif


// Object headers
#include "PrintSubSys.h"

// Message header
#include "dtprintinfomsg.h"
nl_catd dtprintinfo_cat = NULL;

#include <stdlib.h> // This is for the getenv function
#include <unistd.h> // This is for the getuid function
#include <string.h> 
#include <errno.h> 

#if defined(aix) || defined(USL) || defined(__uxp__)
extern "C" { extern int seteuid(uid_t); }
#endif

#ifdef hpux
static char **msg_strings = NULL;

// Cannot use multiple catgets paramemter calls in functions because the 
// previous catgets returned value is overwritten by later catgets calls
// Example: this would fail on HP systems
//    sprintf(buf, "%s %s" catgets(...), catgets(...))

char *Catgets(nl_catd catd, int set_num, int msg_num, char *s)
{
   if (!msg_strings)
      return s;

   if (!msg_strings[msg_num])
      msg_strings[msg_num] = strdup(catgets(catd, set_num, msg_num, s));
   return msg_strings[msg_num];
}
#endif

#ifdef __osf__
extern "C" { extern int svc_init(); }
#endif

int main(int argc, char **argv)
{

#ifdef __osf__
    svc_init();
#endif
  
#ifndef NO_CDE
   _DtEnvControl(DT_ENV_SET);
#endif

// run as user's UID
#ifdef hpux
   setresuid(getuid(), getuid(), (uid_t)0);
#else
   seteuid(getuid());
#endif

   setlocale(LC_ALL, "");

   char *lang = getenv("LANG");
   if (lang && strcmp(lang, "C"))
    {
      errno = 0;

#ifdef NL_CAT_LOCALE
      dtprintinfo_cat = catopen("dtprintinfo", NL_CAT_LOCALE);
#else
      dtprintinfo_cat = catopen("dtprintinfo", 0);
#endif

      if ((nl_catd) errno || dtprintinfo_cat == (nl_catd) -1)
         dtprintinfo_cat = NULL;
#ifdef hpux
      else
         msg_strings = (char **)calloc(LAST_MSG_NO, sizeof(char *));
#endif
    }

   if (argc > 1) {
   if (((!STRCMP(argv[1], "-help")) || (!STRCMP(argv[1], "-h")))
      || ((!STRCMP(argv[1], "-p")) && (argv[2] == NULL)))
    {
      char *output;
      char *cmd = new char [strlen(LIST_QUEUES) + 30];
      sprintf(cmd, "%s | awk '{print \"\\t\", $1}'", LIST_QUEUES);
      Invoke *_thread = new Invoke(cmd, &output);
      printf(MESSAGE(CommandLineHelpL), output);
      printf("\n");
      delete output;
      delete cmd;
      delete _thread;
      return 0;
    }
    }

   char *progname = strrchr(argv[0], '/');
   if (progname)
      progname++;
   else
      progname = argv[0];
   if (!STRCMP(argv[1], "-populate"))
    {
      if (getuid() != 0)
       {
	 fprintf(stderr, MESSAGE(RootUserL), progname, "-populate");
	 fprintf(stderr, "\n");
	 return 1;
       }

      PrintSubSystem *prt = new PrintSubSystem(NULL);
      int n_queues = prt->NumChildren();
      // Get Print Subsystem children, (these are queues)
      Queue **queues = (Queue **)prt->Children();
      int i;
      for (i = 0; i < n_queues; i++)
       {
         DtPrinterIcon *icon = new DtPrinterIcon(NULL, NULL, queues[i],
						 INITIALIZE_PRINTERS);
	 icon->CreateActionFile();
	 delete icon;
       }
      return 0;
    }

   DtApp *app = new DtApp(progname, &argc, argv);
   app->Visible(true);
   app->Run();

   return 0;
}
